consensus/cryptarchia-engine/src/lib.rs

```diff
@@ -252,17 +287,100 @@ where
     pub fn fork_choice(&self) -> Branch<Id> {
         let k = self.config.security_param.get().into();
         let s = self.config.s();
-        Self::maxvalid_bg(self.local_chain.clone(), &self.branches, k, s)
+        Self::maxvalid_bg(self.local_chain, &self.branches, k, s)
     }
 
     pub const fn tip(&self) -> Id {
         self.local_chain.id
     }
 
-    // prune all states deeper than 'depth' with regard to the current
-    // local chain except for states belonging to the local chain
-    pub fn prune_forks(&mut self, _depth: u64) {
-        todo!()
+    pub const fn tip_branch(&self) -> &Branch<Id> {
+        &self.local_chain
+    }
+
+    /// Prune all blocks that are included in forks strictly older than 'depth'
+    /// blocks from the current local chain.
+    ///
+    /// For example, if the tip of the canonical chain is at height 10 (i.e., 11
+    /// blocks long), calling `self.prune_forks(10)` will remove any forks
+    /// stemming from the genesis block, with height `0`, which is the 11th
+    /// block in the past.
+    ///
+    /// This function does not apply any particular logic when evaluating forks
+    /// other than the height at which they diverged from the local
+    /// canonical chain.
+    ///
+    /// It returns the block IDs that were part of the pruned forks.
+    pub fn prune_forks(&mut self, depth: u64) -> impl Iterator<Item = Id> {
+        let prunable_forks = self.prunable_forks(depth);
+        let mut removed_blocks = vec![];
+        for prunable_fork_info in prunable_forks {
+            removed_blocks.extend(self.prune_fork(&prunable_fork_info));
+        }
+        removed_blocks.into_iter()
```

Comment:
I think the low level is ok to force the collects, but here is not necessary at all
```suggestion
    pub fn prune_forks(&mut self, depth: u64) -> impl Iterator<Item = Id> {
        self.prunable_forks(depth).iter().flat_map(|prunable_fork_info| self.prune_fork(&prunable_fork_info))
    }
```
I know you would have to deal with lifetimes probably. If its too difficult we can revisit this later.

Reply:
I gave it a try first: https://github.com/logos-co/nomos/pull/1261/commits/6b3d08bd871fe04571efe0f2f3aac961edde3a05, but then had to re-introduce at least the `collect` part here: https://github.com/logos-co/nomos/pull/1261/commits/4519b7e428101062fbb7f0f7cf40f3d9bc2e0a07.